In [1]:
import pandas as pd
import tensorflow as tf
import numpy as np
import os
import datetime
from tensorflow.keras.callbacks import TensorBoard
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.models import Model
from tensorflow.keras.callbacks import EarlyStopping, TensorBoard
# from ipynb.fs.full.Preprocessing import clean_data
# %load_ext tensorboard

# df = pd.read_csv('./Data/songs_cleaned.csv', index_col='Unnamed: 0')
# print(df.shape)
# df.describe()

In [3]:
df = pd.read_csv('data/songs_cleaned.csv', index_col='Unnamed: 0')
df.head()

popularity  duration_ms  explicit  danceability    energy  key_1  key_2  \
0   -1.153128    -0.765231 -0.218921      0.476586 -0.390370      1      0   
1   -1.480897    -0.978657 -0.218921      0.776863 -1.112757      1      0   
2   -1.480897    -0.358223 -0.218921     -0.790583 -1.454105      0      1   
3   -1.480897    -0.393417 -0.218921     -1.469210 -1.781164      0      0   
4   -1.480897    -0.496230 -0.218921     -0.982761 -1.529520      0      0   

   key_3  key_4  key_5  ...  key_12  loudness  mode  speechiness  \
0      0      0      0  ...       0 -0.593681     1     1.669464   
1      0      0      0  ...       0 -2.308889     1     4.199676   
2      0      0      0  ...       0 -2.122512     1    -0.329704   
3      1      0      0  ...       0 -3.444497     1    -0.333705   
4      0      1      0  ...       0 -1.288108     0    -0.390709   

   acousticness  instrumentalness  liveness   valence     tempo  \
0      0.637973          2.373142 -0.361517 -1.670382 -0.457098   
1      0.990898         -0.422701 -0.377335  0.383636 -0.552432   
2      1.556151         -0.340780 -0.039884 -0.386621  0.400537   
3      1.559021          3.027009 -0.609333 -0.620032  1.727630   
4      1.541805          0.065820  0.482112 -1.401959 -0.511809   

   time_signature  
0       -1.772010  
1       -5.847677  
2        2.303657  
3       -1.772010  
4        0.265824  

[5 rows x 26 columns]

In [4]:
# Simple autoencoder

encoded_dim = 16

input_song = Input(shape = (df.shape[1], ))

h1 = Dense(32, activation = 'relu')(input_song)

encoded = Dense(encoded_dim, activation = 'relu')(h1)

dh1 = Dense(32, activation='sigmoid')(encoded)

decoded = Dense(df.shape[1], activation='sigmoid')(dh1)

autoencoder = Model(input_song, decoded)

encoder = Model(input_song, encoded)

autoencoder.compile(optimizer='nadam', loss='mean_squared_error')

In [5]:
stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=3)
now = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")

logdir = os.path.join("logs", f"SimpleAutoencoder-{now}")
tensorboard = TensorBoard(log_dir=logdir)

autoencoder.fit(df, # input image to encoder
                df, # provide input image to decoder so the model learns how to reconstruct the input image 
                epochs=100,
                validation_split=.2,
                callbacks=[stop, tensorboard])

Train on 356958 samples, validate on 89240 samples
Epoch 1/100
356958/356958 [==============================] - 19s 53us/sample - loss: 0.3449 - val_loss: 0.3302
Epoch 2/100
356958/356958 [==============================] - 18s 49us/sample - loss: 0.3013 - val_loss: 0.3178
Epoch 3/100
356958/356958 [==============================] - 17s 49us/sample - loss: 0.2968 - val_loss: 0.3160
Epoch 4/100
356958/356958 [==============================] - 18s 51us/sample - loss: 0.2952 - val_loss: 0.3147
Epoch 5/100
356958/356958 [==============================] - 21s 60us/sample - loss: 0.2941 - val_loss: 0.3140
Epoch 6/100
356958/356958 [==============================] - 19s 52us/sample - loss: 0.2938 - val_loss: 0.3138
Epoch 7/100
356958/356958 [==============================] - 18s 50us/sample - loss: 0.2936 - val_loss: 0.3136
Epoch 8/100
356958/356958 [==============================] - 18s 51us/sample - loss: 0.2936 - val_loss: 0.3136
Epoch 9/100
356958/356958 [==============================] - 

### Shallow NN Summary:
0 hidden layers: Val Loss: 0.3160
1 hidden layer: 

In [6]:
decoded_songs = autoencoder.predict(df)
encoded_songs = encoder.predict(df)

embeddings_df = pd.DataFrame(encoded_songs, index=df.index)

In [7]:
print(embeddings_df.shape)
embeddings_df

(446198, 16)


0         1         2         3         4         5         6   \
0       3.439245  2.355044  5.307463  3.102348  3.383404  5.992368  1.891921   
1       3.199280  2.001670  5.128112  3.147471  2.665279  4.024767  3.632333   
2       3.591749  4.792721  3.141135  3.261882  6.295177  6.339336  3.863820   
3       5.395530  3.829431  4.966428  1.351473  5.682446  5.453233  2.780163   
4       4.582920  2.215892  1.896022  1.337512  2.313035  3.370011  1.609889   
...          ...       ...       ...       ...       ...       ...       ...   
446470  6.179933  2.420247  6.579400  3.647833  6.442467  6.301361  2.372988   
446471  4.440663  3.161665  2.685592  3.113587  4.283250  4.970417  2.663100   
446472  3.948999  3.675984  2.833682  4.552578  4.047510  4.544388  2.473383   
446473  4.931068  4.872447  3.078540  1.990307  5.198609  3.295120  3.197024   
446474  5.806770  5.723465  3.309666  3.921585  2.775800  4.596958  5.015268   

              7         8         9         10        11        12        13  \
0       3.461351  2.323342  3.324887  5.347540  4.386366  3.943117  3.611048   
1       3.958691  3.334461  2.468137  3.629158  2.196515  2.755299  5.503534   
2       3.113749  5.122262  2.571957  3.807695  4.522155  0.364863  5.790422   
3       4.594506  4.072587  6.494727  4.476008  4.978988  5.414867  4.064033   
4       1.253814  0.336525  1.829196  3.494988  3.029218  1.785699  4.873641   
...          ...       ...       ...       ...       ...       ...       ...   
446470  2.640686  3.161202  1.670727  3.629464  3.976546  4.156880  2.359338   
446471  2.287277  1.904765  2.891276  1.017821  3.729634  2.580269  3.806695   
446472  1.715546  3.869210  2.863277  1.864632  3.380309  4.433583  3.678036   
446473  4.495516  2.935908  3.041452  0.543906  4.423180  4.646929  3.062718   
446474  5.210911  5.211022  3.849360  4.022012  5.496405  7.845721  2.730862   

              14        15  
0       2.821781  7.015693  
1       2.635271  6.248461  
2       1.784324  5.888541  
3       0.430416  5.558275  
4       3.425582  2.965908  
...          ...       ...  
446470  1.559114  7.479818  
446471  4.098092  4.624074  
446472  4.255605  5.757043  
446473  1.828564  4.873153  
446474  3.980950  3.357030  

[446198 rows x 16 columns]

In [8]:
from sklearn.neighbors import NearestNeighbors

nn = NearestNeighbors(n_neighbors=11, algorithm='ball_tree')
nn.fit(encoded_songs)

NearestNeighbors(algorithm='ball_tree', n_neighbors=11)

In [22]:
len(encoded_songs[0])

16

In [ ]:
# Saving the encoded_songs embedding as a df and the Nearest Neighbours model for use in our app
import pickle

embeddings_df = pd.DataFrame(encoded_songs, index=df.index)

# Saving the embeddings to the disk
embeddings_df.to_csv('data/autoencoder_embeddings.csv')

# save the model to disk
filename = 'autoencoder_nn_model.sav'
pickle.dump(nn, open(filename, 'wb'))

# some time later...
 
# load the model from disk
#loaded_model = pickle.load(open(filename, 'rb'))
#result = loaded_model.score(X_test, Y_test)
#print(result)

In [16]:
embeddings_df

0         1    2         3         4         5         6   \
0       3.470329  1.184762 -0.0  3.144354  1.041142  2.506661  2.963388   
1       2.950846  0.692970 -0.0  1.260800  2.372062  1.424065  1.766102   
2       4.472703  6.916587 -0.0  3.051311  4.255427  4.078246  6.610500   
3       0.533942  1.481426 -0.0  4.247745  1.795213  3.724570  6.466702   
4       2.028276  4.151319 -0.0  2.583655  2.820859  2.327025  2.896990   
...          ...       ...  ...       ...       ...       ...       ...   
586667  4.507507  4.525337 -0.0  1.797992  4.408144  3.942313  3.207572   
586668  4.967984  3.997606 -0.0  2.682946  5.384268  4.684574  2.361519   
586669  2.513383  4.620409 -0.0  3.026764  4.275283  6.560318  5.074978   
586670  4.446574  4.300193 -0.0  4.233162  4.264293  5.235272  1.241963   
586671  3.847120  3.737806 -0.0  2.963249  5.151583  2.118279  3.689731   

              7         8         9         10        11        12        13  \
0       1.308488  1.876889  1.609339  5.813118  4.160755  2.801373  6.369508   
1       3.973808  1.751205  1.503640  4.932892  4.102685  1.840484  4.968472   
2       5.424572  5.226394  2.885791  4.807835  8.209723  6.890685  2.668237   
3       2.883682  2.688429  2.568929  3.822564  2.750004  4.210648  3.195352   
4       3.921444  3.261746  1.292208  6.060862  3.102732  4.927852  3.230224   
...          ...       ...       ...       ...       ...       ...       ...   
586667  3.014376  2.259826  3.485083  7.321132  4.829626  4.268849  4.802366   
586668  3.863641  2.298209  4.028028  7.587534  5.696170  5.347732  3.508698   
586669  4.096501  2.595194  1.959881  3.780759  4.040783  5.324980  2.569675   
586670  4.270548  4.267136  2.311171  6.389156  5.381336  5.270690  4.331300   
586671  4.534883  3.179928  3.346431  3.523212  3.196295  6.148501  0.977880   

              14        15  
0       4.479137  2.048349  
1       4.601698  2.108422  
2       4.231436  3.645994  
3       3.263384  3.226213  
4       1.068353  2.544864  
...          ...       ...  
586667  3.543452  2.489765  
586668  5.025582  3.510531  
586669  2.156150  2.363571  
586670  3.387321  5.655649  
586671  3.550117  3.168971  

[586344 rows x 16 columns]

In [23]:
# Testing our model with multiple songs
raw_df = pd.read_csv('./Data/tracks.csv')

test_songs = {'California Love': '0LyLz6XsVs6wz85dK0S6EG', 'The Gambler': '5KqldkCunQ2rWxruMEtGh0', 
              'Eye of the Tiger': '2HHtWyy5CgaQbC7XSoOb0e', 'You are my Sunshine': '4ApTr4wbyOucy2lbAvyuuV'}

# Loop over all our test songs and get the recommendations
recommendations = {}

print('DeepSpot\'s recommendation system:\n\n')

for name, id in test_songs.items():
    query = raw_df.index[raw_df['id'] == id][0]
    test_song = pd.DataFrame([raw_df.iloc[query, :]])
    
    score, close_image_idx = nn.kneighbors(encoded_songs[query].reshape(1,-1))
    k_closest_songs_indexes = close_image_idx[0][1:]
    
    print('Top 10 recommendations for {} by {}:\n'.format(name, test_song.artists))
    for i in range(len(k_closest_songs_indexes)):
        song = raw_df.iloc[k_closest_songs_indexes[i], :]
        print('{} by {} with a score of {}\n'.format(song['name'], song.artists, score[0][1:][i]))
        print('URL: https://open.spotify.com/track/{}\n'.format(song['id']))
    print('___________________________________________________________________________________\n')
    
    recommendations[name] = k_closest_songs_indexes


DeepSpot's recommendation system:


Top 10 recommendations for California Love by 175588    ['2Pac']
Name: artists, dtype: object:

Lisboa Antigua by ['Jorge Sepúlveda'] with a score of 0.0

URL: https://open.spotify.com/track/2TE7NXdEmPnQxi69tAaECa

Hyväillä ja hoivata by ['Frederik'] with a score of 1.0386382294544085

URL: https://open.spotify.com/track/5yV3xS10IBHVkupMLz0ppe

Fade to Black - Live / Seattle '89 by ['Metallica'] with a score of 1.2134997379075512

URL: https://open.spotify.com/track/46GRDe7B3ptpFE7brebD69

Neviem byt sam by ['Elán'] with a score of 1.261382514555003

URL: https://open.spotify.com/track/1l2bM511Nfmoc55dsnCeUp

Heavy Metal: The Black and Silver by ['Blue Öyster Cult'] with a score of 1.2747358413875256

URL: https://open.spotify.com/track/1uUCpOX86yRGShambUp93O

I Remember Jim by ['Anita Kerr Singers'] with a score of 1.2759895467963422

URL: https://open.spotify.com/track/03p1DF2p7e0kRw4UPWDZri

El Sonido De La Ciudad - Remastered by ['Miguel Ríos'] w

In [ ]:
# Deep variational encoder

encoded_dim = 16

input_song = Input(shape = (df.shape[1], ))

encoded_h1 = Dense(64, activation = 'relu')(input_song)
encoded_h2 = Dense(32, activation = 'relu')(encoded_h1)
encoded = Dense(encoded_dim, activation = 'relu')(encoded_h2)

decoded_h1 = Dense(32, activation='relu')(encoded)
decoded_h2 = Dense(64, activation='relu')(decoded_h1)
decoded = Dense(df.shape[1], activation='sigmoid')(decoded_h2)

deep_autoencoder = Model(input_song, decoded) # techincally autoencoder_and_decoder

deep_encoder = Model(input_song, encoded) # Half of the autoencoder

deep_autoencoder.compile(optimizer='nadam', loss='mean_squared_error')

In [ ]:
stop = EarlyStopping(monitor='val_loss', min_delta=0.001, patience=3)
now = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")

logdir = os.path.join("logs", f"SimpleAutoencoder-{now}")
tensorboard = TensorBoard(log_dir=logdir)

deep_autoencoder.fit(df, # input image to encoder
                df, # provide input image to decoder so the model learns how to reconstruct the input image 
                epochs=100,
                validation_split=.2,
                callbacks=[stop, tensorboard])

### Deep NN Summary
Val loss = 0.2960

In [32]:
df.dtypes

popularity          float64
duration_ms         float64
explicit            float64
danceability        float64
energy              float64
key_1                 int64
key_2                 int64
key_3                 int64
key_4                 int64
key_5                 int64
key_6                 int64
key_7                 int64
key_8                 int64
key_9                 int64
key_10                int64
key_11                int64
key_12                int64
loudness            float64
mode                  int64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
time_signature      float64
dtype: object

In [ ]:
decoded_songs = autoencoder.predict(df)
encoded_songs = encoder.predict(df)

from sklearn.neighbors import NearestNeighbors

nn = NearestNeighbors(n_neighbors=11, algorithm='ball_tree')
nn.fit(encoded_songs)

In [ ]:
encoded_songs.shape

In [ ]:
# # Testing
# raw_df = pd.read_csv('./Data/tracks.csv')

# test_songs = {'California Love': '0LyLz6XsVs6wz85dK0S6EG', 'The Gambler': '5KqldkCunQ2rWxruMEtGh0', 
#               'Eye of the Tiger': '2HHtWyy5CgaQbC7XSoOb0e', 'You are my Sunshine': '4ApTr4wbyOucy2lbAvyuuV'}

# # Loop over all our test songs and get the recommendations
# recommendations = {}

# print('DeepSpot\'s recommendation system:\n\n')

# for name, id in test_songs.items():
#     query = raw_df.index[raw_df['id'] == id][0]
#     test_song = pd.DataFrame([raw_df.iloc[query, :]])
    
#     score, close_image_idx = nn.kneighbors(encoded_songs[query].reshape(1,-1))
#     k_closest_songs_indexes = close_image_idx[0][1:]
    
#     print('Top 10 recommendations for {} by {}:\n'.format(name, test_song.artists))
#     for i in range(len(k_closest_songs_indexes)):
#         song = raw_df.iloc[k_closest_songs_indexes[i], :]
#         print('{} by {} with a score of {}\n'.format(song['name'], song.artists, score[0][1:][i]))
#         print('URL: https://open.spotify.com/track/{}\n'.format(song['id']))
#     print('___________________________________________________________________________________\n')
    
#     recommendations[name] = k_closest_songs_indexes



# Training a Variational Autoencoder

In [24]:
# Variational Autoencoder
from keras import layers

class Sampling(layers.Layer):
    """
    Uses (z_mean, z_log_var) to sample a z vector from the latent N-dimensional Normal distribution
    i.e. Z is the vector encoding a digit.
    """
    def call(self, inputs):
        
        # recall from the video that z_mean and z_log_var are vectors 
        z_mean, z_log_var = inputs
        batch = tf.shape(z_mean)[0]
        dim = tf.shape(z_mean)[1]
        
        # sample from an N-dimensional normal distribution
        # epsilon is given shape (batch, dim) because we are adding it to z_mean which has shape (batch, dim)
        epsilon = tf.keras.backend.random_normal(shape=(batch, dim))
        
        # output of tf.exp() is a vector 
        simga = tf.exp(0.5 * z_log_var)
        
        # this is our hideen latent vector made up of a mean and variance vector 
        # variance vector is scaled by epsilon, which is sampled from a normal distribtuion 
        # the the video guy said "stocastic" in reference to epsilon, he meant random 
        Z = z_mean + simga * epsilon
        
        # return hidden latent vector 
        return Z

In [96]:
# recall from the video that the more dimensions that our latent vector has
# the better the results of our model 
latent_dim = 8

# shape of our input data
# we are creating our input layer using Keras's Input() class
# the only thing that input layers really do is define the dimensionality of the input data for the model
encoder_inputs = Input(shape=(df.shape[1], ))

# these are the hidden layers

# pass data vector into FCFF layer 
encoded = Dense(encoded_dim, activation = 'relu')(encoder_inputs)

# recall that ordinarly the output of the last encoding layer is the latent vector 
# but here we are creating two output layers for our encoder - one for the mean and one for the log variance 
# returns a 2-dim mean vector
z_mean = Dense(latent_dim, name="z_mean")(encoded)
# returns a 2-dim log variance vecotr 
z_log_var = Dense(latent_dim, name="z_log_var")(encoded)
# pass mean and variance vector into Sampling class in order to create the Z vector,  Z = mean + var * epsilon
z = Sampling()([z_mean, z_log_var])

# ok, now let's put it all together 
# this is our encoder model 
# inputs are the original images
# outputs are the Z vectors: mean, log variance, and the complete Z, i.e. vector Z = mean + var * epsilon
encoder = Model(inputs=encoder_inputs, outputs=[z_mean, z_log_var, z], name="encoder")
encoder.summary()

Model: "encoder"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_13 (InputLayer)           [(None, 26)]         0                                            
__________________________________________________________________________________________________
dense_22 (Dense)                (None, 16)           432         input_13[0][0]                   
__________________________________________________________________________________________________
z_mean (Dense)                  (None, 8)            136         dense_22[0][0]                   
__________________________________________________________________________________________________
z_log_var (Dense)               (None, 8)            136         dense_22[0][0]                   
____________________________________________________________________________________________

In [97]:
# the input layer to our decoder has the same dimensionality as the latent vector
# because the latent vector is the input to the decoder model
latent_inputs = Input(shape=(latent_dim,))

# these are the hidden layers of our decoder 
# the data at this point is in a vector, the Z latent vector 

# this is the final layer in out decoder
# therefore this layer outputs the reconstruction of the original image 
decoder_outputs = Dense(df.shape[1], activation='sigmoid')(latent_inputs)

# this is our decoder model 
decoder = Model(inputs=latent_inputs, outputs=decoder_outputs, name="decoder")
decoder.summary()

Model: "decoder"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_14 (InputLayer)        [(None, 8)]               0         
_________________________________________________________________
dense_23 (Dense)             (None, 26)                234       
Total params: 234
Trainable params: 234
Non-trainable params: 0
_________________________________________________________________


In [33]:
from tensorflow import keras

class VAE(keras.Model):
    def __init__(self, encoder, decoder, **kwargs):
        """
        This class build a Variational Auto-Encoder. It accepts an Encoder and Decoder model as input. 
        
        Note
        ----
        This VAE class is inheriting Keras's Model API so that it can use the Model class methods 

        """
        # how python 3 handels inheritance 
        super(VAE, self).__init__(**kwargs)
        # set encoder model as class attribute
        self.encoder = encoder
        # set decoder model as class attribute 
        self.decoder = decoder
        # set mean function as class attribute - this calculates the total loss
        self.total_loss_tracker = keras.metrics.Mean(name="total_loss")
        # set mean function as class attribute - this calculates the reconstruction loss
        self.reconstruction_loss_tracker = keras.metrics.Mean(name="reconstruction_loss")
        # set mean function as class attribute - this calculates the kl loss
        self.kl_loss_tracker = keras.metrics.Mean(name="kl_loss")

    @property
    def metrics(self):
        """
        Returns all loses in a list
        """
        return [
            self.total_loss_tracker,
            self.reconstruction_loss_tracker,
            self.kl_loss_tracker,
        ]

    def train_step(self, data):
        """
        Training our model via gradient descent and back-propagation 
        """
        
        # we used tf.GradientTape() in Sprint 2 Module 2 to run Gradient Descent from scratch 
        with tf.GradientTape() as tape:
            # pass input data into encoder model 
            # output of encoder model is the hidden state distribution parameters and hidden state vector 
            z_mean, z_log_var, z = self.encoder(data)
            
            # pass hidden state vector into decoder model 
            reconstruction = self.decoder(z)
            
            # calculate the reconstruction loss 
            reconstruction_loss = tf.reduce_mean(
                tf.reduce_sum(
                    keras.losses.mse(data, reconstruction)#, axis=(1, 2)
                )
            )
            
            # calculate the kl loss
            #                (1 + z_simga   - (z_mean)^2        - e^(z_simga) ) 
            kl_loss = -0.5 * (1 + z_log_var - tf.square(z_mean) - tf.exp(z_log_var))
            # recall that we used tf.reduce_sum() in Sprint 2 Module 4 assignment 
            # it takes the sum of the vector components 
            kl_loss = tf.reduce_mean(tf.reduce_sum(kl_loss, axis=1))
            
            # calculate the total loss by adding the two loss components 
            total_loss = reconstruction_loss + kl_loss
            
        # now that we have calculated the loss function, we can perform Gradient Descent
        # we are passing in the loss function and the weights that we want to update via Gradeint Descent 
        grads = tape.gradient(total_loss, self.trainable_weights)
        self.optimizer.apply_gradients(zip(grads, self.trainable_weights))
        
        # log the total loss
        self.total_loss_tracker.update_state(total_loss)
        
        # log the reconstruction loss 
        self.reconstruction_loss_tracker.update_state(reconstruction_loss)
        
        # log the kl loss 
        self.kl_loss_tracker.update_state(kl_loss)
        
        # return all the losses in a dictionary 
        return {
            "loss": self.total_loss_tracker.result(),
            "reconstruction_loss": self.reconstruction_loss_tracker.result(),
            "kl_loss": self.kl_loss_tracker.result(),
        }

In [98]:
# instantiate a Variational Auto-Encoder model 
vae = VAE(encoder, decoder)

# complie the model 
vae.compile(optimizer=keras.optimizers.Nadam())

stop = EarlyStopping(monitor='loss', min_delta=0.001, patience=3, restore_best_weights=True)
now = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")

logdir = os.path.join("logs", f"VAE-{now}")
tensorboard = TensorBoard(log_dir=logdir)

# train the model weights 
vae.fit(df, epochs=100,
        callbacks=[stop, tensorboard])

2021-08-25 10:37:44.631468: I tensorflow/core/profiler/lib/profiler_session.cc:126] Profiler session initializing.
2021-08-25 10:37:44.631485: I tensorflow/core/profiler/lib/profiler_session.cc:141] Profiler session started.
2021-08-25 10:37:44.631659: I tensorflow/core/profiler/lib/profiler_session.cc:159] Profiler session tear down.


Epoch 1/100
  125/18324 [..............................] - ETA: 30s - loss: 25.0390 - reconstruction_loss: 23.1630 - kl_loss: 1.2593

2021-08-25 10:37:45.749862: I tensorflow/core/profiler/lib/profiler_session.cc:126] Profiler session initializing.
2021-08-25 10:37:45.749879: I tensorflow/core/profiler/lib/profiler_session.cc:141] Profiler session started.
2021-08-25 10:37:45.810550: I tensorflow/core/profiler/lib/profiler_session.cc:66] Profiler session collecting data.
2021-08-25 10:37:45.812944: I tensorflow/core/profiler/lib/profiler_session.cc:159] Profiler session tear down.
2021-08-25 10:37:45.815861: I tensorflow/core/profiler/rpc/client/save_profile.cc:137] Creating directory: logs/VAE-20210825-103744/train/plugins/profile/2021_08_25_10_37_45
2021-08-25 10:37:45.817346: I tensorflow/core/profiler/rpc/client/save_profile.cc:143] Dumped gzipped tool data for trace.json.gz to logs/VAE-20210825-103744/train/plugins/profile/2021_08_25_10_37_45/Petrs-MBP.trace.json.gz
2021-08-25 10:37:45.821984: I tensorflow/core/profiler/rpc/client/save_profile.cc:137] Creating directory: logs/VAE-20210825-103744/train/plugins/pr

18324/18324 [==============================] - 16s 859us/step - loss: 17.4896 - reconstruction_loss: 14.8143 - kl_loss: 1.56901s - loss: 17.6129 - reconstruc
Epoch 2/100
18324/18324 [==============================] - 14s 745us/step - loss: 15.6907 - reconstruction_loss: 13.9358 - kl_loss: 1.7640
Epoch 3/100
18324/18324 [==============================] - 14s 750us/step - loss: 15.6536 - reconstruction_loss: 13.8345 - kl_loss: 1.8127
Epoch 4/100
18324/18324 [==============================] - 14s 748us/step - loss: 15.6581 - reconstruction_loss: 13.7940 - kl_loss: 1.8307
Epoch 5/100
18324/18324 [==============================] - 14s 758us/step - loss: 15.6480 - reconstruction_loss: 13.7708 - kl_loss: 1.8459
Epoch 6/100
18324/18324 [==============================] - 14s 759us/step - loss: 15.5487 - reconstruction_loss: 13.7483 - kl_loss: 1.8573
Epoch 7/100
18324/18324 [==============================] - 14s 747us/step - loss: 15.6579 - reconstruction_loss: 13.7274 - kl_loss: 1.8741
Epoch 8/

# VAE Summary:
loss: 15.5321 - reconstruction_loss: 13.6563 - kl_loss: 1.9151

In [51]:
# Encoding our songs
z_mean, z_log_var, z = vae.encoder.predict(df)

embeddings_df = pd.DataFrame(z, index=df.index)

# fitting KNN

nn = NearestNeighbors(n_neighbors=11, algorithm='ball_tree')
nn.fit(z)

NearestNeighbors(algorithm='ball_tree', leaf_size=30, metric='minkowski',
                 metric_params=None, n_jobs=None, n_neighbors=11, p=2,
                 radius=1.0)

In [66]:
# Testing the VAE - It seems to work!
raw_df = pd.read_csv('./Data/tracks.csv')

test_songs = {'California Love': '0LyLz6XsVs6wz85dK0S6EG', 'The Gambler': '5KqldkCunQ2rWxruMEtGh0', 
              'Eye of the Tiger': '2HHtWyy5CgaQbC7XSoOb0e', 'You are my Sunshine': '4ApTr4wbyOucy2lbAvyuuV'}

# Loop over all our test songs and get the recommendations
recommendations = {}

print('DeepSpot\'s recommendation system:\n\n')

for name, id in test_songs.items():
    query = raw_df.index[raw_df['id'] == id][0]
    test_song = pd.DataFrame([raw_df.iloc[query, :]])
    
    score, close_image_idx = nn.kneighbors(z[query].reshape(1,-1))
    k_closest_songs_indexes = close_image_idx[0][1:]
    
    print('Top 10 recommendations for {} by {}:\n'.format(name, test_song.artists))
    for i in range(len(k_closest_songs_indexes)):
        song = raw_df.iloc[k_closest_songs_indexes[i], :]
        print('{} by {} with a score of {}\n'.format(song['name'], song.artists, score[0][1:][i]))
        print('URL: https://open.spotify.com/track/{}\n'.format(song['id']))
    print('___________________________________________________________________________________\n')
    
    recommendations[name] = k_closest_songs_indexes




DeepSpot's recommendation system:


Top 10 recommendations for California Love by 175588    ['2Pac']
Name: artists, dtype: object:

Torpedo Blu by ['Giorgio Gaber'] with a score of 2.0387225534784523

URL: https://open.spotify.com/track/2hWsjxSO7bIKHzCBQMmmA0

I Am a Rocker - Live at Abbey Road; 2005 Remaster by ['Chas & Dave'] with a score of 2.042183201276265

URL: https://open.spotify.com/track/1rNfjUa8wtQNeg5CpnZrpw

September '99 - Phats & Small Remix by ['Earth, Wind & Fire', 'Phats & Small'] with a score of 2.0909346043421233

URL: https://open.spotify.com/track/6Z6my7Cpi5BongU2fsY96l

Mera Yaar by ['Shankar-Ehsaan-Loy', 'Javed Bashir'] with a score of 2.0985039762456625

URL: https://open.spotify.com/track/2NECp5Tvb5Lm6TbJf0Y8xI

Tracks of My Tears by ['Smokey Robinson'] with a score of 2.1003578809715298

URL: https://open.spotify.com/track/5QizJ1IfhrJemVys14KlJF

It Don't Matter to Me by ['Bread'] with a score of 2.104974467345315

URL: https://open.spotify.com/track/5s8lZd5T

# Hyperparameter tuning the VAE

In [82]:
latent_dim = 8

encoder_inputs = Input(shape=(df.shape[1], ))

h1 = Dense(16, activation = 'relu')(encoder_inputs)
encoded = Dense(encoded_dim, activation = 'relu')(h1)

z_mean = Dense(latent_dim, name="z_mean")(encoded)

z_log_var = Dense(latent_dim, name="z_log_var")(encoded)

z = Sampling()([z_mean, z_log_var])

encoder = Model(inputs=encoder_inputs, outputs=[z_mean, z_log_var, z], name="encoder")
encoder.summary()

Model: "encoder"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_11 (InputLayer)           [(None, 26)]         0                                            
__________________________________________________________________________________________________
dense_18 (Dense)                (None, 16)           432         input_11[0][0]                   
__________________________________________________________________________________________________
dense_19 (Dense)                (None, 16)           272         dense_18[0][0]                   
__________________________________________________________________________________________________
z_mean (Dense)                  (None, 8)            136         dense_19[0][0]                   
____________________________________________________________________________________________

In [83]:
latent_inputs = Input(shape=(latent_dim,))

dh1 = Dense(16, activation='relu')(latent_inputs)

decoder_outputs = Dense(df.shape[1], activation='sigmoid')(dh1)

# this is our decoder model 
decoder = Model(inputs=latent_inputs, outputs=decoder_outputs, name="decoder")
decoder.summary()

Model: "decoder"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_12 (InputLayer)        [(None, 8)]               0         
_________________________________________________________________
dense_20 (Dense)             (None, 16)                144       
_________________________________________________________________
dense_21 (Dense)             (None, 26)                442       
Total params: 586
Trainable params: 586
Non-trainable params: 0
_________________________________________________________________


In [84]:
# instantiate a Variational Auto-Encoder model 
vae = VAE(encoder, decoder)

# complie the model 
vae.compile(optimizer=keras.optimizers.Nadam())

stop = EarlyStopping(monitor='loss', min_delta=0.001, patience=3, restore_best_weights=True)
now = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")

logdir = os.path.join("logs", f"DeepVAE-{now}")
tensorboard = TensorBoard(log_dir=logdir)

# train the model weights 
vae.fit(df, epochs=100,
        callbacks=[stop, tensorboard])

2021-08-25 10:26:16.779240: I tensorflow/core/profiler/lib/profiler_session.cc:126] Profiler session initializing.
2021-08-25 10:26:16.779260: I tensorflow/core/profiler/lib/profiler_session.cc:141] Profiler session started.
2021-08-25 10:26:16.779510: I tensorflow/core/profiler/lib/profiler_session.cc:159] Profiler session tear down.


Epoch 1/100
  105/18324 [..............................] - ETA: 41s - loss: 25.9393 - reconstruction_loss: 23.2566 - kl_loss: 0.9896 

2021-08-25 10:26:18.450155: I tensorflow/core/profiler/lib/profiler_session.cc:126] Profiler session initializing.
2021-08-25 10:26:18.450169: I tensorflow/core/profiler/lib/profiler_session.cc:141] Profiler session started.
2021-08-25 10:26:18.522889: I tensorflow/core/profiler/lib/profiler_session.cc:66] Profiler session collecting data.
2021-08-25 10:26:18.525891: I tensorflow/core/profiler/lib/profiler_session.cc:159] Profiler session tear down.
2021-08-25 10:26:18.529263: I tensorflow/core/profiler/rpc/client/save_profile.cc:137] Creating directory: logs/DeepVAE-20210825-102616/train/plugins/profile/2021_08_25_10_26_18
2021-08-25 10:26:18.531174: I tensorflow/core/profiler/rpc/client/save_profile.cc:143] Dumped gzipped tool data for trace.json.gz to logs/DeepVAE-20210825-102616/train/plugins/profile/2021_08_25_10_26_18/Petrs-MBP.trace.json.gz
2021-08-25 10:26:18.537788: I tensorflow/core/profiler/rpc/client/save_profile.cc:137] Creating directory: logs/DeepVAE-20210825-102616/trai

18324/18324 [==============================] - 17s 880us/step - loss: 17.0531 - reconstruction_loss: 15.2413 - kl_loss: 1.0129
Epoch 2/100
18324/18324 [==============================] - 15s 833us/step - loss: 15.6979 - reconstruction_loss: 14.2567 - kl_loss: 1.4092
Epoch 3/100
18324/18324 [==============================] - 16s 849us/step - loss: 15.6211 - reconstruction_loss: 14.1369 - kl_loss: 1.4779
Epoch 4/100
18324/18324 [==============================] - 16s 893us/step - loss: 15.5848 - reconstruction_loss: 14.0539 - kl_loss: 1.5281
Epoch 5/100
18324/18324 [==============================] - 18s 964us/step - loss: 15.5693 - reconstruction_loss: 14.0180 - kl_loss: 1.5495
Epoch 6/100
18324/18324 [==============================] - 16s 898us/step - loss: 15.5970 - reconstruction_loss: 14.0010 - kl_loss: 1.5576
Epoch 7/100
18324/18324 [==============================] - 16s 866us/step - loss: 15.4935 - reconstruction_loss: 13.9793 - kl_loss: 1.5698
Epoch 8/100
18324/18324 [==============

Results deep VAE v1 (latent_dim = 16, h1 = 32): 
loss: 15.9362 - reconstruction_loss: 14.5233 - kl_loss: 1.4098

**Results deep VAE v2 (latent_dim = 8, h1 = 16):**
Epoch 21/100
loss: 15.4746 - reconstruction_loss: 13.8923 - kl_loss: 1.6215

Results regular VAE v1 (latent_dim = 16): 
loss: 15.5321 - reconstruction_loss: 13.6563 - kl_loss: 1.9151

Results regular VAE v1 (latent_dim = 8):
Epoch 12/100
loss: 15.5430 - reconstruction_loss: 13.7091 - kl_loss: 1.8810

In [95]:
# Testing results
# Encoding our songs
z_mean, z_log_var, z = vae.encoder.predict(df)

embeddings_df = pd.DataFrame(z, index=df.index)

# fitting KNN

nn = NearestNeighbors(n_neighbors=11, algorithm='ball_tree')
nn.fit(z)

# Testing the VAE v2
raw_df = pd.read_csv('./Data/tracks.csv')

test_songs = {'California Love': '0LyLz6XsVs6wz85dK0S6EG', 'The Gambler': '5KqldkCunQ2rWxruMEtGh0', 
              'Eye of the Tiger': '2HHtWyy5CgaQbC7XSoOb0e', 'You are my Sunshine': '4ApTr4wbyOucy2lbAvyuuV'}

# Loop over all our test songs and get the recommendations
recommendations = {}

print('DeepSpot\'s recommendation system:\n\n')

for name, id in test_songs.items():
    query = raw_df.index[raw_df['id'] == id][0]
    test_song = pd.DataFrame([raw_df.iloc[query, :]])
    
    score, close_image_idx = nn.kneighbors(z[query].reshape(1,-1))
    k_closest_songs_indexes = close_image_idx[0][1:]
    
    print('Top 10 recommendations for {} by {}:\n'.format(name, test_song.artists))
    for i in range(len(k_closest_songs_indexes)):
        song = raw_df.iloc[k_closest_songs_indexes[i], :]
        print('{} by {} with a score of {}\n'.format(song['name'], song.artists, score[0][1:][i]))
        print('URL: https://open.spotify.com/track/{}\n'.format(song['id']))
    print('___________________________________________________________________________________\n')
    
    recommendations[name] = k_closest_songs_indexes

embeddings_df.to_csv('./Data/vae_embeddings.csv')



DeepSpot's recommendation system:


Top 10 recommendations for California Love by 175588    ['2Pac']
Name: artists, dtype: object:

Psycho (feat. Ty Dolla $ign) by ['Post Malone', 'Ty Dolla $ign'] with a score of 0.5826315098088156

URL: https://open.spotify.com/track/3swc6WTsr7rl9DqQKQA55C

Han Mikkel by ['Unit Five'] with a score of 0.6039448575863513

URL: https://open.spotify.com/track/0SSDgzZ20kSjzRMsx4CXTo

Verdi: Rigoletto, Act 2: "Tutte le feste al tempio" (Rigoletto, Gilda) by ['Giuseppe Verdi', 'Tito Gobbi', 'Orchestra Del Teatro Alla Scala, Milano', 'Tullio Serafin'] with a score of 0.6166760292215672

URL: https://open.spotify.com/track/1rsvdwKNP7r5cRjPf74Yw9

สัญญาเจ้าลืม by ['หนู มิเตอร์'] with a score of 0.6176218466138796

URL: https://open.spotify.com/track/1IzLBR8lmdKmmagDAzolA6

Pissikset by ['Lissut'] with a score of 0.6213709595482744

URL: https://open.spotify.com/track/2R22hS7NsTueQrG7z4Fq5d

Gaucho by ['Steely Dan'] with a score of 0.6272687563675855

URL: https

In [ ]:
# Save the entire model as a SavedModel.
!mkdir -p saved_model
autoencoder.save('saved_model/vae_002')
encoder.save('saved_model/encoder_002')